File: docforge/packages/doc-core/src/hash.ts
/**
 * 稳定 content-hash。浏览器/Node 通用(纯 JS,无 node:crypto 依赖)。
 * 用于编译/预览缓存:相同 IR -> 相同 hash -> 命中缓存跳过编译。
 */

/** 稳定序列化:对象键排序,保证语义相同的 IR 产出相同字符串。 */
export function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value) ?? "null";
  }
  if (Array.isArray(value)) {
    return `[${value.map(stableStringify).join(",")}]`;
  }
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  const parts = keys
    .filter((k) => obj[k] !== undefined)
    .map((k) => `${JSON.stringify(k)}:${stableStringify(obj[k])}`);
  return `{${parts.join(",")}}`;
}

/** FNV-1a 64-bit (用 BigInt),返回 16 位十六进制字符串。 */
export function fnv1a64(str: string): string {
  const FNV_OFFSET = 0xcbf29ce484222325n;
  const FNV_PRIME = 0x100000001b3n;
  const MASK = 0xffffffffffffffffn;
  let hash = FNV_OFFSET;
  for (let i = 0; i < str.length; i++) {
    hash ^= BigInt(str.charCodeAt(i) & 0xff);
    hash = (hash * FNV_PRIME) & MASK;
    // 处理多字节字符的高位
    const hi = str.charCodeAt(i) >> 8;
    if (hi) {
      hash ^= BigInt(hi);
      hash = (hash * FNV_PRIME) & MASK;
    }
  }
  return hash.toString(16).padStart(16, "0");
}

/** 对任意 IR / 值计算稳定哈希。 */
export function contentHash(value: unknown): string {
  return fnv1a64(stableStringify(value));
}

File: rust/crates/moonlit-core/src/hash.rs
//! Stable content hashing that is byte-for-byte compatible with the JS
//! implementation in `docforge/packages/doc-core/src/hash.ts`.
//!
//! Used as the compile/preview cache key: identical IR -> identical hash ->
//! cache hit. Parity with the JS hash is therefore a hard requirement; the
//! parity tests in `moonlit-doccore` depend on it.

use serde_json::Value;

/// Stable serialization: object keys are sorted so that semantically equal
/// values produce identical strings. Mirrors `stableStringify` in hash.ts.
///
/// Note: JS `JSON.stringify` renders integral numbers without a trailing
/// `.0` (e.g. `1`, not `1.0`). We replicate that here so the FNV input
/// matches the JS string exactly.
pub fn stable_stringify(value: &Value) -> String {
    match value {
        Value::Null => "null".to_string(),
        Value::Bool(b) => {
            if *b {
                "true".to_string()
            } else {
                "false".to_string()
            }
        }
        Value::Number(n) => format_number(n),
        Value::String(_) => {
            // serde_json escapes strings the same way as JS JSON.stringify
            // for the BMP/standard cases used by the IR.
            serde_json::to_string(value).unwrap_or_else(|_| "null".to_string())
        }
        Value::Array(items) => {
            let parts: Vec<String> = items.iter().map(stable_stringify).collect();
            format!("[{}]", parts.join(","))
        }
        Value::Object(map) => {
            // JS only filters `undefined`; serde_json has no `undefined`, so we
            // keep every present key (including explicit null).
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            let parts: Vec<String> = keys
                .into_iter()
                .map(|k| {
                    let key = serde_json::to_string(&Value::String(k.clone()))
                        .unwrap_or_else(|_| format!("\"{k}\""));
                    format!("{}:{}", key, stable_stringify(&map[k]))
                })
                .collect();
            format!("{{{}}}", parts.join(","))
        }
    }
}

/// Render a JSON number the way `JSON.stringify` would in JS: integral floats
/// drop the fractional part.
fn format_number(n: &serde_json::Number) -> String {
    if let Some(i) = n.as_i64() {
        return i.to_string();
    }
    if let Some(u) = n.as_u64() {
        return u.to_string();
    }
    if let Some(f) = n.as_f64() {
        if f.is_finite() && f.fract() == 0.0 && f.abs() < 1e15 {
            return format!("{}", f as i64);
        }
        // Shortest round-trippable representation, matching V8 closely enough
        // for the coordinate magnitudes used by the IR.
        let s = format!("{f}");
        return s;
    }
    // NaN / Infinity are serialized as null by JSON.stringify.
    "null".to_string()
}

/// FNV-1a 64-bit over the UTF-16 code units of `s`, returning a zero-padded
/// 16-char hex string. Mirrors `fnv1a64` in hash.ts (which iterates over
/// `charCodeAt`, i.e. UTF-16 code units, low byte then optional high byte).
pub fn fnv1a64(s: &str) -> String {
    const FNV_OFFSET: u64 = 0xcbf2_9ce4_8422_2325;
    const FNV_PRIME: u64 = 0x0000_0100_0000_01b3;
    let mut hash: u64 = FNV_OFFSET;
    for unit in s.encode_utf16() {
        let lo = (unit & 0xff) as u64;
        hash ^= lo;
        hash = hash.wrapping_mul(FNV_PRIME);
        let hi = unit >> 8;
        if hi != 0 {
            hash ^= hi as u64;
            hash = hash.wrapping_mul(FNV_PRIME);
        }
    }
    format!("{hash:016x}")
}

/// Stable content hash over any JSON value.
pub fn content_hash(value: &Value) -> String {
    fnv1a64(&stable_stringify(value))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn stable_stringify_sorts_keys() {
        let v = json!({ "b": 1, "a": 2 });
        assert_eq!(stable_stringify(&v), "{\"a\":2,\"b\":1}");
    }

    #[test]
    fn stable_stringify_integral_numbers_have_no_decimal() {
        let v = json!({ "x": 1.0, "w": 8.0 });
        assert_eq!(stable_stringify(&v), "{\"w\":8,\"x\":1}");
    }

    #[test]
    fn stable_stringify_fractional_numbers_kept() {
        let v = json!({ "h": 5.625 });
        assert_eq!(stable_stringify(&v), "{\"h\":5.625}");
    }

    #[test]
    fn fnv_matches_known_vectors() {
        // Empty string -> FNV offset basis (16 hex).
        assert_eq!(fnv1a64(""), "cbf29ce484222325");
        // "a": offset ^ 0x61 then * prime.
        let expected = {
            let mut h: u64 = 0xcbf2_9ce4_8422_2325;
            h ^= 0x61;
            h = h.wrapping_mul(0x0000_0100_0000_01b3);
            format!("{h:016x}")
        };
        assert_eq!(fnv1a64("a"), expected);
    }

    #[test]
    fn content_hash_is_stable_across_key_order() {
        let a = json!({ "type": "word", "blocks": [] });
        let b = json!({ "blocks": [], "type": "word" });
        assert_eq!(content_hash(&a), content_hash(&b));
    }
}
